### ISSUE TITLE
Incorrect result of _MKLMaxPoolGrad

### ISSUE THREAD

**Issue by djshen** [tensorflow/tensorflow#40122]
**System information**
- OS Platform and Distribution: `Arch Linux 5.5.2-arch1-1  x86_64`
- TensorFlow installed from: `source`
- TensorFlow version: `v1.12.1-33097-g83eb4048ba 2.2.0` and `v2.2.0-0-g2b96f3662b 2.2.0`
- Python version: `Python 3.6.10`
- Bazel version: `3.0.0` for master, `2.0.0` for r2.2
- GCC/Compiler version: `GCC 9.3.0`

The package was built with the commands:
```bash
bazel build --config=mkl //tensorflow/tools/pip_package:build_pip_package
# For master (commit #83eb40)
./bazel-bin/tensorflow/tools/pip_package/build_pip_package --nightly_flag ./master-83eb40
# For r2.2
bazel build --config=mkl //tensorflow/tools/pip_package:build_pip_package
```

**Describe the current behavior**
The gradient of the max pooling 2D is wrong.

Code:
```python
import numpy as np
import tensorflow as tf
tf.compat.v1.disable_v2_behavior()

x = np.array([
    [3, 0, 0, 2, 3],
    [0, 0, 0, 0, 1],
    [0, 0, 0, 1, 3],
    [0, 0, 0, 0, 0],
    [1, 1, 3, 8, 6]
]).astype(np.float32).reshape([1, 5, 5, 1])

x_t = tf.compat.v1.placeholder(tf.float32, shape=[1, 5, 5, 1])
w = np.array([1]).reshape([1, 1, 1, 1]).astype(np.float32)
conv_t = tf.nn.conv2d(x_t, w, [1, 1, 1, 1], 'SAME')
pool_t = tf.nn.max_pool(conv_t, [1, 2, 2, 1], [1, 2, 2, 1], 'VALID')
grad_t = tf.gradients(ys=pool_t, xs=conv_t)

tensors = [conv_t, pool_t, grad_t]
tensors = [tf.squeeze(t, [-1]) for t in tensors]

with tf.compat.v1.Session() as sess:
    conv, pool, grad = sess.run(tensors, feed_dict={x_t: x})
    print('conv\n', conv, '\npool\n', pool, '\ngrad\n', grad)
```
Output:
```
conv
 [[[3. 0. 0. 2. 3.]
  [0. 0. 0. 0. 1.]
  [0. 0. 0. 1. 3.]
  [0. 0. 0. 0. 0.]
  [1. 1. 3. 8. 6.]]]
pool
 [[[3. 2.]
  [0. 1.]]]
grad
 [[[[1. 0. 1. 0. 0.]
   [0. 0. 0. 0. 0.]
   [1. 0. 1. 0. 0.]
   [0. 0. 0. 0. 0.]
   [0. 0. 0. 0. 0.]]]]
```

**Describe the expected behavior**
If we run the code with `TF_DISABLE_MKL=1`, the gradient will be
```
 [[[[1. 0. 0. 1. 0.]
   [0. 0. 0. 0. 0.]
   [1. 0. 0. 1. 0.]
   [0. 0. 0. 0. 0.]
   [0. 0. 0. 0. 0.]]]]
```
Note that the positions of the second `1`'s in the first and the third rows are different.

**Other info / logs**
If I directly feed the input to max pooling, the result is correct.
```python
import numpy as np
import tensorflow as tf
tf.compat.v1.disable_v2_behavior()


x = np.array([
    [3, 0, 0, 2, 3],
    [0, 0, 0, 0, 1],
    [0, 0, 0, 1, 3],
    [0, 0, 0, 0, 0],
    [1, 1, 3, 8, 6]
]).astype(np.float32).reshape([1, 5, 5, 1])

x_t = tf.compat.v1.placeholder(tf.float32, shape=[1, 5, 5, 1])
pool_t = tf.nn.max_pool(x_t, [1, 2, 2, 1], [1, 2, 2, 1], 'VALID')
grad_t = tf.gradients(ys=pool_t, xs=x_t)

tensors = [x_t, pool_t, grad_t]
tensors = [tf.squeeze(t, [-1]) for t in tensors]

with tf.compat.v1.Session() as sess:
    x, pool, grad = sess.run(tensors, feed_dict={x_t: x})
    print('x\n', x, '\npool\n', pool, '\ngrad\n', grad)
```
Output:
```
x
 [[[3. 0. 0. 2. 3.]
  [0. 0. 0. 0. 1.]
  [0. 0. 0. 1. 3.]
  [0. 0. 0. 0. 0.]
  [1. 1. 3. 8. 6.]]]
pool
 [[[3. 2.]
  [0. 1.]]]
grad
 [[[[1. 0. 0. 1. 0.]
   [0. 0. 0. 0. 0.]
   [1. 0. 0. 1. 0.]
   [0. 0. 0. 0. 0.]
   [0. 0. 0. 0. 0.]]]]
```
In addition, if the I replace the conv2d with relu, which will be also rewritten, the result is also correct.
```python
import numpy as np
import tensorflow as tf
tf.compat.v1.disable_v2_behavior()


x = np.array([
    [3, 0, 0, 2, 3],
    [0, 0, 0, 0, 1],
    [0, 0, 0, 1, 3],
    [0, 0, 0, 0, 0],
    [1, 1, 3, 8, 6]
]).astype(np.float32).reshape([1, 5, 5, 1])

x_t = tf.compat.v1.placeholder(tf.float32, shape=[1, 5, 5, 1])
relu_t = tf.nn.relu(x_t)
pool_t = tf.nn.max_pool(relu_t, [1, 2, 2, 1], [1, 2, 2, 1], 'VALID')
grad_t = tf.gradients(ys=pool_t, xs=relu_t)

tensors = [relu_t, pool_t, grad_t]
tensors = [tf.squeeze(t, [-1]) for t in tensors]

with tf.compat.v1.Session() as sess:
    relu, pool, grad = sess.run(tensors, feed_dict={x_t: x})
    print('relu\n', relu, '\npool\n', pool, '\ngrad\n', grad)
```
Output:
```
relu
 [[[3. 0. 0. 2. 3.]
  [0. 0. 0. 0. 1.]
  [0. 0. 0. 1. 3.]
  [0. 0. 0. 0. 0.]
  [1. 1. 3. 8. 6.]]]
pool
 [[[3. 2.]
  [0. 1.]]]
grad
 [[[[1. 0. 0. 1. 0.]
   [0. 0. 0. 0. 0.]
   [1. 0. 0. 1. 0.]
   [0. 0. 0. 0. 0.]
   [0. 0. 0. 0. 0.]]]]
```
I checked the log with `TF_CPP_MIN_VLOG_LEVEL=1` and confirmed that the OP was rewritten.

It seems that the result is affected by the convolution.

**Comment by tabaddor** [comment]
I'd like to look into this issue

**Comment by ravikyram** [comment]
@djshen 

I have tried in colab with TF version 2.2 and i am not seeing any issue here.Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/b89f9c38e637d8507935e6aeb6fb8a88/untitled47.ipynb).Thanks!

**Comment by djshen** [comment]
@ravikyram 

I think the official release is not compiled with MKL. This issue does not exist when I install TF from PyPI.

**Comment by TensorFlow-MKL** [comment]
we are looking at this issue and will respond once we reproduce

**Comment by NeoZhangJianyu** [comment]
@djshen We are checking this issue now.

**Comment by NeoZhangJianyu** [comment]
@djshen @ravikyram @ymodak @pre
I reproduce the issue in TF 2.2 (Installed by Conda: tensorflow  2.2.0   mkl_py36h5a57954_0).
There is no such problem in TF2.1.

TF2.2 use DNNL 1.2.2 release.
After check the DNNL log, it's the fault of DNNL (MKLDNN). 
I have reported the issue to Intel DNNL team as high level.

Hope get the confirm and fix as soon!

**Comment by NeoZhangJianyu** [comment]
DNNL team confirm it's not DNNL issue. I have reported to Intel Tensorflow team as high level.

**Comment by nammbash** [comment]
@penpornk Internally we are also working to solve this bug. Just a heads up. No milestone set for TF 2.3 yet. But making a mention for your internal tracking.

**Comment by penpornk** [comment]
@nammbash Got it. Thank you!

**Comment by NeoZhangJianyu** [comment]
@djshen 

This issue is fixed in Tf 2.4.

Could you check it?

**Comment by djshen** [comment]
@NeoZhangJianyu 

I compiled TF v2.4.0-rc0 from source with `--config=mkl` and confirmed that the result is correct.

**Comment by NeoZhangJianyu** [comment]
@djshen 
It's great news!

Could you close this issue? :)

**Comment by djshen** [comment]
@NeoZhangJianyu 
Sure!

**Comment by NeoZhangJianyu** [comment]
@djshen 

Thank you very much!
